File: algorithmic-toolbox/week2/solutions/6_last_digit_of_the_sum_of_fibonacci_numbers/python/fibonacci_sum_last_digit.py
def fib_pisano_period(m):
    pisano_period = [0, 1]
    prev, current = 0, 1
    while True:
        current, prev = (current + prev) % m, current
        if current == 1 and prev == 0:
            return pisano_period[:-1]
        pisano_period.append(current)


def fibonacci_sum(n):
    pisano_period = fib_pisano_period(10)
    pisano_period_sum = sum(pisano_period)
    return ((pisano_period_sum * (n // 60)) + sum(pisano_period[: (n % 60) + 1])) % 10


if __name__ == "__main__":
    n = int(input())
    print(fibonacci_sum(n))

File: algorithmic-toolbox/week2/solutions/6_last_digit_of_the_sum_of_fibonacci_numbers/cpp/fibonacci_sum_last_digit.cpp
#include <iostream>

long long fibonacci_sum(long long n)
{
    if (n <= 1)
    {
        return n;
    }
    long long prev = 0;
    long long current = 1;
    long long swap = 0;
    long long sum = 1;
    for (long long i = 1; i < n; ++i)
    {
        swap = current;
        current = (current + prev) % 10;
        prev = swap;
        sum = sum + current;
    }
    return sum % 10;
}

int main()
{
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum(n);
}
